### DISCUSSION TITLE
OpenAPI specification for CAPI

### DISCUSSION THREAD

**Discussion by koyadume** [kubernetes-sigs/cluster-api#6481] [Q&A] [answered]
I started looking into the documentation of CAPI recently. I was expecting to see a link to OpenAPI specification for it in the documentation assuming that it is an API at a minimum but could not find any. 

It seems to be a collection of tools (clusterctl etc) so why exactly it is called an API? If it is an API too, where can I find OpenAPI specification for it?

**Comment by schrej** [comment, marked as answer]
It is an API. The API is not served by Cluster API but uses the CustomResourceDefinition mechanism of the Kubernetes API server to provide it.

The CustomResourceDefinitions, which are a full description of the API partially consisting of openAPIv3 spec, reside [here](https://github.com/kubernetes-sigs/cluster-api/tree/main/config/crd/bases) and are generated from [these Go types](https://github.com/kubernetes-sigs/cluster-api/tree/main/api).
There is a useful tool that renders the types (or CRD definitions, not sure actually) into a nice website: https://doc.crds.dev/github.com/kubernetes-sigs/cluster-api

**Reply by koyadume** [reply]
Thanks for the explanation.

**Reply by fabriziopandini** [reply]
FYI, in the context of the RuntimeSDK, we are considering generating a new release artifact with the openAPIv3 spec for the core types
@sbueringer @ykakarap @killianmuldoon

**Reply by sbueringer** [reply]
I thought this was for the hooks only? What would the exact format for CRDs look like? Is there prior art?

**Reply by schrej** [reply]
You could probably just export them from the kube-apiserver. Apparently there are openapi endpoints that serve the spec for all known resources: https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-specification

**Reply by sbueringer** [reply]
Thx, good point. I wasn't aware that the apiserver now also publishes the v3 specifications.

I'm not sure if we should really add additional release artifacts for the OpenAPI v3 specs of our CRDs given that we already publish the CRDs as part of our manifests. (except if there's a good use case to publish the standalone specifications as well)
